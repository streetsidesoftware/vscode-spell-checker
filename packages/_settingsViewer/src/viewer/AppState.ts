/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import { observable, computed, makeObservable, action } from 'mobx';
import { Settings, ConfigTarget, LocaleId, SettingByConfigTarget, WorkspaceFolder, TextDocument, ConfigSource } from '../api/settings/';
import { normalizeCode, lookupCode } from '../iso639-1';
import { compareBy, compareEach } from '../api/utils/Comparable';
import { uniqueFilter } from '../api/utils';
import { Messenger } from '../api';
import { ConfigTargets, configTargets } from '../api/settings/settingsHelper';

type TabTargets = ConfigTarget | 'file' | 'dictionaries' | 'about';

export interface Tab {
    label: string;
    target: TabTargets;
}

const tabs: Tab[] = [
    { label: 'User', target: 'user' },
    { label: 'Workspace', target: 'workspace' },
    { label: 'Folder', target: 'folder' },
    { label: 'File', target: 'file' },
    { label: 'Dictionaries', target: 'dictionaries' },
    { label: 'About', target: 'about' },
];

export interface LanguageInfo {
    code: string;
    name: string;
    dictionaries: string[];
    enabled: boolean;
}

export interface LanguageConfig {
    languages: LanguageInfo[];
    inherited?: ConfigSource;
}

export interface LanguageConfigs extends SettingByConfigTarget<LanguageConfig> {}

export interface State {
    activeTabName: string;
    settings: Settings;
    tabs: Tab[];
    activeTab: Tab;
    languageConfig: LanguageConfigs;
}

export interface FoundInConfig<T> {
    value: Exclude<T, undefined>;
    source: ConfigSource;
}

type InheritedFromSource<T> = {
    value: Exclude<T, undefined>;
    source: ConfigSource;
};

export class AppState implements State {
    @observable _activeTabName = '';
    @observable _settings: Settings = {
        dictionaries: [],
        knownLanguageIds: [],
        configs: {
            user: { inherited: {}, locales: [], languageIdsEnabled: [] },
            workspace: { inherited: { locales: 'user', languageIdsEnabled: 'user' }, locales: [], languageIdsEnabled: [] },
            folder: { inherited: { locales: 'user', languageIdsEnabled: 'user' }, locales: [], languageIdsEnabled: [] },
            file: undefined,
        },
    };
    @observable debugMode: boolean = false;

    constructor(private messageBus: Messenger) {
        makeObservable(this);
    }

    @computed get activeTabName() {
        return this._activeTabName;
    }

    @computed get settings() {
        return this._settings;
    }

    @computed get tabs() {
        const hidden = new Set<TabTargets>(configTargets.filter((target) => !this.settings.configs[target]));
        if (this.workspaceFolders.length <= 1) {
            hidden.add(ConfigTargets.folder);
        }
        if (this.workspaceFolders.length < 1) {
            hidden.add(ConfigTargets.workspace);
        }
        if (!this.settings.activeFileUri) {
            hidden.add('file');
        }
        return tabs.filter((tab) => !hidden.has(tab.target));
    }

    @computed get activeTab() {
        return this.tabs.find((t) => t.label === this.activeTabName) || this.tabs[0];
    }

    @computed get languageConfig(): LanguageConfigs {
        const calcConfig = (target: ConfigTarget): LanguageConfig => {
            const config = this.settings.configs[target];
            const locales = config.locales; // todo: calc inheritance
            const inherited = config.inherited.locales;

            const infos = new Map<string, LanguageInfo>();

            const addLocalesToInfos = (locales: string[], dictionaryName: string | undefined) => {
                locales
                    .map(normalizeCode)
                    .map(lookupCode)
                    .filter(notUndefined)
                    .forEach((lang) => {
                        const { code, lang: language, country } = lang;
                        const name = country ? `${language} - ${country}` : language;
                        const found = this.isLocalEnabledEx(target, code);
                        const enabled = (found && found.value) || false;
                        const info: LanguageInfo = infos.get(name) || {
                            code,
                            name,
                            enabled,
                            dictionaries: [],
                        };
                        if (dictionaryName) {
                            info.dictionaries.push(dictionaryName);
                        }
                        infos.set(name, info);
                    });
            };
            if (locales) {
                addLocalesToInfos(locales, undefined);
            }
            this.settings.dictionaries.forEach((dict) => addLocalesToInfos(dict.locales, dict.name));

            return {
                languages: [...infos.values()].sort(
                    compareEach(
                        compareBy((info) => !info.dictionaries.length),
                        compareBy('name')
                    )
                ),
                inherited,
            };
        };

        return {
            user: calcConfig('user'),
            workspace: calcConfig('workspace'),
            folder: calcConfig('folder'),
        };
    }

    @computed get activeTabIndex(): number {
        const index = this.tabs.findIndex((t) => t.label === this.activeTabName);
        return index > 0 ? index : 0;
    }

    @computed get workspaceFolderNames(): string[] {
        const workspace = this.settings.workspace;
        const folders = (workspace && workspace.workspaceFolders) || [];
        return folders.map((f) => f.name);
    }

    @computed get activeWorkspaceFolder(): string | undefined {
        const folder = this.findMatchingFolder(this.settings.activeFolderUri);
        return folder && folder.name;
    }

    @computed get workspaceFolders(): WorkspaceFolder[] {
        const workspace = this.settings.workspace;
        return (workspace && workspace.workspaceFolders) || [];
    }

    @computed get activeFileUri(): string | undefined {
        return this.settings.activeFileUri;
    }

    @computed get documents(): TextDocument[] {
        const workspace = this.settings.workspace;
        return (workspace && workspace.textDocuments) || [];
    }

    @computed get documentSelection(): { label: string; value: string }[] {
        return this.documents.map((doc) => ({ label: doc.fileName, value: doc.uri }));
    }

    private findMatchingFolder(uri: string | undefined): WorkspaceFolder | undefined {
        return this.workspaceFolders.filter((f) => f.uri === uri)[0];
    }

    private findMatchingFolderByName(name: string | undefined): WorkspaceFolder | undefined {
        return this.workspaceFolders.filter((f) => f.name === name)[0];
    }

    @action actionSetLocale(target: ConfigTarget, locale: LocaleId, enable: boolean) {
        const localesCurrent = this.settings.configs[target].locales;
        const locales = (enable ? [...localesCurrent, locale] : localesCurrent.filter((v) => v !== locale)).filter(uniqueFilter());
        this.settings.configs[target].locales = locales;
        const uri = this.settings.activeFolderUri;
        this.messageBus.postMessage({ command: 'EnableLocaleMessage', value: { target, locale, enable, uri } });
    }

    @action actionSetDebugMode(isEnabled: boolean) {
        this.debugMode = isEnabled;
    }

    private isLocalEnabledEx(field: ConfigTarget, code: LocaleId): InheritedFromSource<boolean> {
        const config = this.settings.configs[field];
        const source = config.inherited.locales || field;
        const locales = config.locales;
        return {
            value: locales.map(normalizeCode).includes(code),
            source,
        };
    }

    @action actionActivateTabIndex(index: number) {
        const tab = this.tabs[index];
        if (tab) {
            this.actionActivateTab(tab.label);
        }
    }

    @action actionActivateTab(tabName: string) {
        this._activeTabName = tabName;
        this.messageBus.postMessage({ command: 'SelectTabMessage', value: this.activeTabName });
    }

    @action actionSelectFolder(folderName: string) {
        const folder = this.findMatchingFolderByName(folderName);
        const folderUri = folder && folder.uri;
        this.settings.activeFolderUri = folderUri;
        if (folderUri !== undefined) {
            this.messageBus.postMessage({ command: 'SelectFolderMessage', value: folderUri });
        }
    }

    @action actionSelectDocument(documentUri: string) {
        this.messageBus.postMessage({ command: 'SelectFileMessage', value: documentUri });
    }

    @action actionEnableLanguageId(languageId: string, enable: boolean, target?: ConfigTarget) {
        const fileConfig = this.settings.configs.file;
        if (!target && fileConfig?.languageId === languageId) {
            fileConfig.languageEnabled = enable;
        }
        const uri = (!target && fileConfig?.uri) || this.settings.activeFolderUri;
        this.messageBus.postMessage({ command: 'EnableLanguageIdMessage', value: { languageId, enable, target, uri } });
    }

    @action updateSettings(settings: Settings): Settings {
        this._settings = settings;
        return this._settings;
    }
}

function notUndefined<T>(a: T): a is Exclude<T, undefined> {
    return a !== undefined;
}
