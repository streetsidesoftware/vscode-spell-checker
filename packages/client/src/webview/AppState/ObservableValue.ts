import deepEqual from 'fast-deep-equal';
import { createDisposable, type DisposableHybrid as Disposable } from 'utils-disposables';

export interface ObservableValue<T> {
    value: T;
    set(v: T): void;
    update(u: (v: T) => T): void;
    subscribe(s: (v: T) => unknown): Disposable;
}
class Observable<T> implements ObservableValue<T> {
    private _value: T;
    private _subscriptions = new Set<(v: T) => unknown>();
    private _busy = false;
    constructor(value: T) {
        this._value = value;
    }

    get value() {
        return this._value;
    }

    set value(v: T) {
        this.set(v);
    }

    set(value: T) {
        // Do not update if the value has not changed.
        if (this._value === value || deepEqual(this._value, value)) return;
        this._value = value;
        this.notify();
        return;
    }

    subscribe(s: (v: T) => unknown): Disposable {
        const subscriptions = this._subscriptions;
        subscriptions.add(s);
        return createDisposable(() => subscriptions.delete(s));
    }

    update(u: (v: T) => T) {
        return this.set(u(this._value));
    }

    private notify() {
        if (this._busy) return;
        try {
            this._busy = true;
            for (const s of this._subscriptions) {
                s(this._value);
            }
        } finally {
            this._busy = false;
        }
    }
}
export function createStoreValue<T>(v: T): ObservableValue<T> {
    return new Observable(v);
}
