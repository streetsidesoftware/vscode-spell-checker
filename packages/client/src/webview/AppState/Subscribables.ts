import deepEqual from 'fast-deep-equal';
import type { DisposableHybrid as Disposable, DisposableLike, DisposeFn } from 'utils-disposables';
import { createDisposable, disposeOf, InheritableDisposable } from 'utils-disposables';

export type MakeSubscribable<T, RO extends keyof T | never = never> = {
    [K in keyof T]: K extends RO ? SubscribableValue<T[K]> : ObservableValue<T[K]>;
};

export type SubscriberFn<T> = (v: T) => void;

export interface Subscribable<T> {
    subscribe(s: SubscriberFn<T>): Disposable;
    dispose?: () => void;
}

export interface SubscribableValue<T> extends Required<Subscribable<T>>, Disposable {
    value: T | undefined;
}

export interface ObservableValue<T> extends SubscribableValue<T> {
    value: T;
    set(v: T): void;
    update(u: (v: T) => T): void;
}

class Observable<T> extends InheritableDisposable implements ObservableValue<T> {
    private _value: T;
    private _subscriptions = new Set<SubscriberFn<T>>();
    private _busy = false;
    constructor(value: T) {
        super();
        this._value = value;
    }

    get value() {
        return this._value;
    }

    set value(v: T) {
        this.set(v);
    }

    set(value: T) {
        // Do not update if the value has not changed.
        if (this._value === value || deepEqual(this._value, value)) return;
        this._value = value;
        this.notify();
        return;
    }

    subscribe(s: (v: T) => unknown): Disposable {
        const subscriptions = this._subscriptions;
        subscriptions.add(s);
        s(this._value);
        return createDisposable(() => subscriptions.delete(s));
    }

    update(u: (v: T) => T) {
        return this.set(u(this._value));
    }

    private notify() {
        if (this._busy) return;
        try {
            this._busy = true;
            for (const s of this._subscriptions) {
                s(this._value);
            }
        } finally {
            this._busy = false;
        }
    }
}

export function createStoreValue<T>(v: T): ObservableValue<T> {
    return new Observable(v);
}

export type SubscribeFn<T> = (emitter: (v: T) => void) => DisposableLike | DisposeFn | undefined;

const symbolNotSet = Symbol('A values that has not been set.');

interface CreateSubscribableValueOptions {
    timeout?: number;
}

export function createSubscribableValue<T>(
    subscribe: SubscribeFn<T>,
    initialValue?: T,
    options?: CreateSubscribableValueOptions,
): SubscribableValue<T> {
    return new SubscribableValueImpl(subscribe, initialValue, options);
}

type TimeOutHandle = ReturnType<typeof setTimeout>;

class SubscribableValueImpl<T> extends InheritableDisposable implements SubscribableValue<T> {
    private _timeout = 5000;
    private _handleTimeout: TimeOutHandle | undefined = undefined;
    private _subscriptions = new Set<SubscriberFn<T>>();
    private _source: SubscribeFn<T>;

    protected _started = false;
    protected _value: T | typeof symbolNotSet;
    protected _dispose: DisposableLike | DisposeFn | undefined;
    protected _busy = false;

    constructor(subscribe: SubscribeFn<T>, initialValue?: T, options?: CreateSubscribableValueOptions) {
        super([() => this._cleanup()]);
        this._source = subscribe;
        this._value = initialValue !== undefined ? initialValue : symbolNotSet;
        this._timeout = options?.timeout ?? this._timeout;
    }

    private _cleanup() {
        this._subscriptions.clear();
        disposeOf(this._dispose);
        this._setTimeout(undefined);
    }

    private _setTimeout(h: TimeOutHandle | undefined) {
        const handle = this._handleTimeout;
        this._handleTimeout = h;
        if (handle) {
            clearTimeout(handle);
        }
    }

    private _stop() {
        if (this._subscriptions.size) return;

        this._started = false;
        disposeOf(this._dispose);
    }

    private _unSub(s: SubscriberFn<T>) {
        this._subscriptions.delete(s);
        if (!this._subscriptions.size) {
            this._setTimeout(setTimeout(() => this._stop(), this._timeout));
        }
    }

    private _start() {
        if (this._started) return;
        this._started = true;
        this._dispose = this._source((v: T) => this._notify(v));
    }

    private _notify(v: T) {
        this._value = v;
        if (this._busy) return;
        try {
            this._busy = true;
            for (const s of this._subscriptions) {
                s(v);
            }
        } finally {
            this._busy = false;
        }
    }

    get value() {
        return this._value === symbolNotSet ? undefined : this._value;
    }

    private _sub(s: SubscriberFn<T>): Disposable {
        this._subscriptions.add(s);
        if (this._value !== symbolNotSet) {
            s(this._value);
        }
        this._start();
        return createDisposable(() => this._unSub(s));
    }

    public subscribe = this._sub.bind(this);
}

export function awaitForSubscribable<T>(sub: Subscribable<T>): Promise<T> {
    let disposable: DisposableLike | undefined;

    return new Promise<T>((resolve, _reject) => {
        // prevent double resolve.
        let resolved = false;
        disposable = sub.subscribe((v) => {
            if (resolved) return;
            resolved = true;
            try {
                resolve(v);
            } finally {
                disposeOf(disposable);
            }
        });
    });
}

export interface Emitter<T> extends Subscribable<T> {
    emit(value: T): void;
    dispose(): void;
}

export function createEmitter<T>(): Emitter<T> {
    const subscriptions = new Set<SubscriberFn<T>>();
    let _busy = false;

    function emit(value: T) {
        if (_busy) return;
        try {
            _busy = true;
            for (const s of subscriptions) {
                s(value);
            }
        } finally {
            _busy = false;
        }
    }

    return {
        emit,
        subscribe(s) {
            subscriptions.add(s);
            return createDisposable(() => subscriptions.delete(s));
        },
        dispose() {
            subscriptions.clear();
        },
    };
}
