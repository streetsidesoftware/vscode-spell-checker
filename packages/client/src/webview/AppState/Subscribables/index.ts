export * from './createFunctions';
export * from './operators/index';
export { pipe } from './pipe';
export { rx } from './rx';
export type * from './Subscribables';
export { createSubscribableView } from './SubscribableView';
