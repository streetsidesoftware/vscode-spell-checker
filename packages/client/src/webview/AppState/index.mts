export { awaitSubscribable } from '../../Subscribables/helpers/awaitSubscribable.js';
export { calcDocSettings, getWebviewGlobalStore } from './store.mjs';
