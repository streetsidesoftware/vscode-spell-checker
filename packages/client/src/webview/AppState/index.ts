export { awaitSubscribable } from '../../Subscribables/helpers/awaitSubscribable';
export { calcDocSettings, getWebviewGlobalStore } from './store';
