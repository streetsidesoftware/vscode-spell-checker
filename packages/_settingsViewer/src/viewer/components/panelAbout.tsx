import * as React from 'react';
import { observer } from 'mobx-react';
import { AppState } from '../AppState';
import { VsCodeWebviewApi } from '../../api/vscode/VsCodeWebviewApi';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction';
import IconSettings from '@material-ui/icons/Settings';
import { CsCheckBox, CsButton, CsList as List } from './primitives';

// resources
import spellCheckIcon from '../images/SpellCheck.xs.png';

const vsCodeApi = new VsCodeWebviewApi();

@observer
export class PanelAbout extends React.Component<{ appState: AppState }> {
    render(): JSX.Element {
        const appState = this.props.appState;
        const workspace = appState.settings.workspace || {};
        const toggle = () => this.props.appState.actionSetDebugMode(!appState.debugMode);
        return (
            <div>
                <h1>
                    <img style={{ verticalAlign: 'middle', paddingBottom: '8px' }} src={spellCheckIcon} />
                    <span>&nbsp;Code Spell Checker</span>
                </h1>

                <CsButton variant="contained" color="primary" onClick={this.onReset}>
                    Refresh
                </CsButton>
                <h2>Options</h2>
                <div>
                    <div>
                        <List>
                            <ListItem onClick={toggle}>
                                <ListItemIcon>
                                    <IconSettings />
                                </ListItemIcon>
                                <ListItemText primary={'Debug Mode'} />
                                <ListItemSecondaryAction>
                                    <CsCheckBox checked={appState.debugMode} onChange={toggle} />
                                </ListItemSecondaryAction>
                            </ListItem>
                        </List>
                    </div>
                </div>
                {/* <DevTools /> */}
                {appState.debugMode ? (
                    <div>
                        <h2>Workspace</h2>
                        <pre>{JSON.stringify(workspace, null, 2)}</pre>

                        <h2>Configs</h2>
                        <pre>{JSON.stringify(appState.settings.configs, null, 2)}</pre>
                    </div>
                ) : (
                    ''
                )}
            </div>
        );
    }

    onReset: () => void = () => {
        vsCodeApi.postMessage({ command: 'RequestConfigurationMessage' });
    };
}
