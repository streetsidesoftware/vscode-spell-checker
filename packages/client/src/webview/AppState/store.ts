import type { DisposableClassic, DisposableHybrid, DisposableLike } from 'utils-disposables';
import { createDisposable, createDisposableFromList, disposeOf, injectDisposable } from 'utils-disposables';
import { LogLevelMasks } from 'utils-logger';
import type { TextDocument, TextEditor, Uri } from 'vscode';
import { window } from 'vscode';
import type { WatchFieldList, WatchFields } from 'webview-api';

import { getDependencies } from '../../di';
import { calcSettings } from '../../infoViewer/infoHelper';
import { createSubscribableView, pipe, rx, throttle } from '../../Subscribables';
import { toSubscriberFn } from '../../Subscribables/helpers/toSubscriber';
import type { MakeSubscribable, StoreValue } from '../../Subscribables/StoreValue';
import { createStoreValue } from '../../Subscribables/StoreValue';
import type { SubscriberLike } from '../../Subscribables/Subscribables';
import { getLogger } from '../api/api';
import type { AppStateData } from '../apiTypes';

export interface Storage {
    seq: number;
    state: MakeSubscribable<AppStateData, 'currentDocument'>;
    dispose(): void;
}

const debug = false;

debug && getLogger().setLogLevelMask(LogLevelMasks.everything);

let store: Storage | undefined = undefined;

export function getWebviewGlobalStore(): Storage {
    if (store) return store;

    const currentDocumentSub = rx(subscribeToCurrentDocument);
    const currentDocument = pipe(currentDocumentSub, throttle(500), /* delayUnsubscribe(5000), */ createSubscribableView);
    // currentDocument.onEvent('onNotify', (event) => console.log('current document update: %o', event));

    function dispose() {
        disposeOf(currentDocumentSub);
        const _store = store;
        store = undefined;
        if (!_store) return;
        Object.values(_store.state).forEach((s) => disposeOf(s));
    }

    const writableState = {
        logDebug: createStoreValue(getLogger().isMethodEnabled('debug')),
        todos: createStoreValue<AppStateData['todos']>([]),
    } as const;

    const _store: Storage = injectDisposable(
        {
            seq: 1,
            state: {
                ...writableState,
                currentDocument,
            },
        },
        dispose,
        'getWebviewGlobalStore',
    );

    return (store = _store);
}

function subscribeToCurrentDocument(subscriber: SubscriberLike<AppStateData['currentDocument']>): DisposableHybrid {
    const emitter = toSubscriberFn(subscriber);
    const disposables: DisposableLike[] = [];
    const disposable = createDisposableFromList(disposables);

    setCurrentDocument(window.activeTextEditor);
    disposables.push(disposeClassic(window.onDidChangeActiveTextEditor(setCurrentDocument, undefined)));
    disposables.push(
        disposeClassic(
            window.onDidChangeTextEditorSelection(
                (event) => event.textEditor === window.activeTextEditor && setCurrentDocument(event.textEditor),
                undefined,
            ),
        ),
    );

    return disposable;

    function setCurrentDocument(textEditor: TextEditor | undefined) {
        if (!textEditor) {
            // emitter(null);
            return;
        }

        const document = textEditor.document;
        const doc = {
            url: document.uri.toString(),
            version: document.version,
        };

        emitter(doc);
    }
}

export async function calcDocSettings(doc?: string) {
    const textDoc = (doc && findMatchTextDocument(doc)) || undefined;
    const di = getDependencies();
    return calcSettings(textDoc, undefined, di.client, console.log);
}

export interface StateUpdate<T> {
    seq: number;
    success: boolean;
    value: T;
}

export function updateState<T>(seq: number | undefined, value: T, s: StoreValue<T>): StateUpdate<T> {
    const store = getWebviewGlobalStore();
    if (seq && seq !== store.seq) return { seq: store.seq, value: s.value, success: false };

    store.seq++;
    s.set(value);
    return { seq: store.seq, value: s.value, success: true };
}

export function watchFieldList(fieldsToWatch: Set<WatchFields>, onChange: (changedFields: WatchFieldList) => void): DisposableHybrid {
    const store = getWebviewGlobalStore();
    const list = [...fieldsToWatch];
    const disposables = list
        .map((field) => ({ field, sub: store.state[field] }))
        .map((ss) => {
            return ss.sub.subscribe(() => onChange([ss.field]));
        });

    return createDisposableFromList(disposables);
}

function findMatchTextDocument(url: UrlLike): TextDocument | undefined {
    return findMatchingEditor(url)?.document;
}

function findMatchingEditor(url: UrlLike): TextEditor | undefined {
    for (const editor of window.visibleTextEditors) {
        if (!compareUrl(editor.document.uri, url)) return editor;
    }
    return undefined;
}

type UrlLike = URL | Uri | string;

function compareUrl(a: UrlLike, b: UrlLike): number {
    const aa = normalizeUrlToString(a);
    const bb = normalizeUrlToString(b);
    if (aa === bb) return 0;
    return aa < bb ? -1 : 1;
}

function normalizeUrlToString(url: UrlLike): string {
    const decoded = decodeURIComponent(decodeURIComponent(url.toString())).normalize('NFC');
    return decoded.replace(/^file:\/\/\/[a-z]:/i, (fileUrl) => fileUrl.toLowerCase());
}

function disposeClassic(disposable: DisposableClassic): DisposableHybrid {
    return createDisposable(() => disposable.dispose(), undefined, 'disposeClassic');
}
