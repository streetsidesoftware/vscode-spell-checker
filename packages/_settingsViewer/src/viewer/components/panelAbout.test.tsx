import * as React from 'react';
import { AppState } from '../AppState';
import { create } from 'react-test-renderer';
import { sampleAppState } from '../../test/fixtures/AppState';
import { PanelAbout } from './panelAbout';

describe('About Panel Verification', () => {
    it('tests the snapshot', () => {
        const appState = getSampleAppState();
        const panelRenderer = create(<PanelAbout appState={appState}></PanelAbout>).toJSON()!;
        expect(panelRenderer).toMatchSnapshot();
    });

    function getSampleAppState(): AppState {
        return sampleAppState();
    }
});
