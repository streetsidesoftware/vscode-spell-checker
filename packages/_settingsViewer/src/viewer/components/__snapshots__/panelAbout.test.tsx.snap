// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`About Panel Verification > tests the snapshot 1`] = `
<div>
  <h1>
    <img
      src="/src/viewer/images/SpellCheck.xs.png"
      style={
        {
          "paddingBottom": "8px",
          "verticalAlign": "middle",
        }
      }
    />
    <span>
      Â Code Spell Checker
    </span>
  </h1>
  <button
    className="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-1 MuiButton-contained MuiButton-containedPrimary WithStyles(ForwardRef(Button))-containedPrimary-2"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Refresh
    </span>
  </button>
  <h2>
    Options
  </h2>
  <div>
    <div>
      <ul
        className="MuiList-root WithStyles(ForwardRef(List))-root-4 MuiList-padding"
      >
        <li
          className="MuiListItem-container"
        >
          <div
            className="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
            disabled={false}
            onClick={[Function]}
          >
            <div
              className="MuiListItemIcon-root"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61l-2.01-1.58zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6z"
                />
              </svg>
            </div>
            <div
              className="MuiListItemText-root"
            >
              <span
                className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Debug Mode
              </span>
            </div>
          </div>
          <div
            className="MuiListItemSecondaryAction-root"
          >
            <span
              aria-disabled={false}
              className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-10 MuiCheckbox-root WithStyles(ForwardRef(Checkbox))-root-5 MuiCheckbox-colorSecondary WithStyles(ForwardRef(Checkbox))-colorSecondary-7 MuiIconButton-colorSecondary"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label"
              >
                <input
                  checked={false}
                  className="PrivateSwitchBase-input-13"
                  data-indeterminate={false}
                  onChange={[Function]}
                  type="checkbox"
                />
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </span>
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
  
</div>
`;
